File: visitor_pattern/C/src/quick_sort.cpp
#include <inc/quick_sort.h>
#include <algorithm>
#include <cstdlib>
template<typename T, typename C>
size_t quick_sort<T,C>::partition(size_t s, size_t e)
{
    C c;
    /* a[s] is the pivot element */
    size_t i = s, j = s;   
    while(++j < e)
    {
        if (c(a[s],a[j]))
            continue; 
        std::swap(a[++i], a[j]);
    }
    std::swap(a[i], a[s]);
    return i;
}
template<typename T, typename C>
void quick_sort<T,C>::sort(size_t s,  size_t e)
{
    if (s + 1 >= e)
        return;
    /* randomize pivot element */
    size_t p = rand() % ( e - s) + s;
    std::swap(a[p], a[0]);
    p = partition(s,e);
    quick_sort(s, p);
    quick_sort(p + 1, e);
}
template<typename T, typename C>
void quick_sort<T,C>::do_work(void)
{
    sort(0, sz);
}

File: visitor_pattern/java/QuickSort.java
//package com.charlesq.java.visitorpattern;
import java.util.Random;
import java.lang.Object;
import java.lang.Comparable;
public class QuickSort<T extends Comparable<T>> implements Visitable<T>
{
     private transient T[] a;
     @Override
     public void  accept(Visitor<T> v)
     {
         v.visit(this);
     }    
     public void init(T [] a)
     {
         this.a = a;
     }
     public void run()
     {
         sort(0, a.length);
     }
     
     /* partition elements in [s, e) */
     private int partition(int s, int e)
     {
        int i = s, j = s;  
        while(++j < e)
        {
     
            if (a[s].compareTo(a[j]) >= 0) 
                swap(++i, j);
        }
        swap(i, s);
        return i;
     }
     public void sort(int s, int e)
     {
         /* termination condition in recusive call, 
            minimal two elements are required */
         if (s + 1 >= e)
             return;
         /* pick a random pivot element */
         int p = (new Random()).nextInt(e-s) + s; 
         swap(p, s);/* start element is the pivot */
         p = partition(s,e);
         sort(s, p); 
         sort(p + 1, e);
     }
     public String toString()
     {
         return "Quick Sort";
     }
     /* A common utility method for sorting algorithm */
     void swap(int i, int j)
     {
         if (i == j)
             return;
         T temp = a[i];
         a[i] = a[j];
         a[j] = temp;
     }

}
